lib/Sema/ConstraintSystem.cpp

```diff
@@ -3696,6 +3706,20 @@ Type ConstraintSystem::simplifyTypeImpl(Type type,
     if (auto tvt = dyn_cast<TypeVariableType>(type.getPointer()))
       return getFixedTypeFn(tvt);
 
+    if (auto tuple = dyn_cast<TupleType>(type.getPointer())) {
+      if (tuple->getNumElements() == 1) {
+        auto element = tuple->getElement(0);
+        auto elementType = simplifyTypeImpl(element.getType(), getFixedTypeFn);
+
+        // Flatten single-element tuples containing type variables that cannot
+        // bind to packs.
+        auto typeVar = elementType->getAs<TypeVariableType>();
+        if (!element.hasName() && typeVar && !typeVar->getImpl().canBindToPack()) {
```

Comment:
This worries me because there could be a case where we have a type variable that can bind to a pack, and the pack binding later on contains a single element, but I haven't thought of a better way to handle this yet.

Reply:
Hm, if I remember correctly `TupleType::get` for a single unlabled element would always produce a paren type instead, does `(repeat each T)` from your example have to be a tuple type because `T` could be a pack expansion?

Reply:
That behavior of TupleType::get() was removed when I introduced one-element tuples in the first place :)

Reply:
So we broke source compatibility in some cases then?

Reply:
I would hope not, since one-element tuples are gated behind the experimental VariadicGenerics feature flag, and I updated all existing callers to explicitly handle the unlabeled one-element case if they expect to get a ParenType back.

Reply:
@slavapestov This works in 5.7:

```
func test() -> (_: Int) { 42 }

let _: Int = test()
```

but it doesn't work with `-enable-experimental-feature -Xfrontend VariadicGenerics` on main.

Reply:
Yeah, I guess in this case we broke source compatibility. Does anyone use this syntax though? Regardless, Holly is removing one-element tuples so this code will once again behave like it did in 5.7.

Reply:
What I wonder is whether we would have issues in places where splats or implicit injections into tuples which used to behave like parens but would start behaving like tuples.

Reply:
I think I'm going to restore the behavior of `TupleType::get` where it can produce a `ParenType` or a `TupleType`, but I'll make that explicit in the return type so callers are forced to handle it. I'll do that in a follow-up PR.

Reply:
Sounds good to me!